libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1411 @@
+                                                     GCancellable  *cancellable,
+                                                     GError       **error)
+{
+  g_autoptr(GHashTable) ref_actions_for_files = g_hash_table_new_full (g_str_hash,
+                                                                       g_str_equal,
+                                                                       g_free,
+                                                                       (GDestroyNotify) flatpak_remote_ref_actions_file_free);
+
+  if (!eos_updater_util_flatpak_ref_actions_append_from_directory (relative_parent_path,
+                                                                   directory,
+                                                                   ref_actions_for_files,
+                                                                   priority,
+                                                                   cancellable,
+                                                                   error))
+    return NULL;
+
+  return g_steal_pointer (&ref_actions_for_files);
+}
+
+static guint
+flatpak_ref_hash (gconstpointer data)
+{
+  return g_str_hash (flatpak_ref_get_name (FLATPAK_REF (data)));
+}
+
+static gboolean
+flatpak_ref_equal (gconstpointer a, gconstpointer b)
+{
+  return g_str_equal (flatpak_ref_get_name (FLATPAK_REF (a)),
+                      flatpak_ref_get_name (FLATPAK_REF (b)));
+}
+
+/* Squash actions on the same ref into the last action on that ref, returning
+ * a pointer array of remote ref actions, ordered by the order key in each
+ * remote ref action. */
+static GPtrArray *
+squash_ref_actions_ptr_array (GPtrArray *ref_actions)
+{
+  g_autoptr(GHashTable) hash_table = g_hash_table_new_full (flatpak_ref_hash,
+                                                            flatpak_ref_equal,
+                                                            g_object_unref,
+                                                            (GDestroyNotify) flatpak_remote_ref_action_unref);
+  GPtrArray *squashed_ref_actions = NULL;
+  gsize i;
+  gpointer key;
+  gpointer value;
+  GHashTableIter hash_iter;
+
+  for (i = 0; i < ref_actions->len; ++i)
+    {
+      FlatpakRemoteRefAction *action = g_ptr_array_index (ref_actions, i);
+      g_hash_table_replace (hash_table,
+                            g_object_ref (action->ref),
+                            flatpak_remote_ref_action_ref (action));
+    }
+
+  squashed_ref_actions = g_ptr_array_new_full (g_hash_table_size (hash_table),
+                                               (GDestroyNotify) flatpak_remote_ref_action_unref);
+  g_hash_table_iter_init (&hash_iter, hash_table);
+
+  while (g_hash_table_iter_next (&hash_iter, &key, &value))
+    g_ptr_array_add (squashed_ref_actions,
+                     flatpak_remote_ref_action_ref (value));
+
+  g_ptr_array_sort (squashed_ref_actions, sort_flatpak_remote_ref_actions);
```

Comment:
Should we be sorting here? By this point, the serial numbers don’t form a total order, since we’re dealing with entries from multiple files (serial number domains). I don’t think you can take a round-trip through the hash table: you’d need to build `squashed_ref_actions` from the original `ref_actions` with a lookup in the hash table for each entry to check whether it’s the latest for that ref (and hence whether to include it).

Reply:
See my comment below - the sorting here is what should be done based on the current design doc.